### PR TITLE
Binary-pairing Coulomb collisions: improvements and optimization

### PR DESCRIPTION
**Pull request by JustinRayAngus**
**Description updated by @RemiLehe, using text by @JustinRayAngus:** 

This PR now just makes some improvements on the relativistic implementation of Coulomb scattering method by writing the formula for the normalized mean free path (s12) in a way that is directly analogous to the well-known non-relativistic formulation. This save flops by removing redundant calculations and improves on the readability of the Coulomb collision algorithm.

The expression for the normalized scattering path in UpdateMomentumPerez.H is written as:
`s12 = sigma_eff * n12 * dt * vrelst * g1sg2s/(g1g2)`
with `sigma_eff = pi*b0^2 * lnLmd`

The proof that this expression is equivalent to Eq. 9 of Perez et al 2012 (https://doi.org/10.1063/1.4742167) is given here 
[CoulombScattering_s12.pdf](https://github.com/user-attachments/files/16663359/CoulombScattering_s12.pdf)

**Previous description:**

This PR originally addressed several issues with the binary-pairing Coulomb collision method in WarpX.
1) Optimization. See PR #5066

2) Weighted-particle scattering physics. See PR #5091. The method for doing weighted-particle scattering for Coulomb collisions prior to PR #5091 is from Perez et al 2012 (https://doi.org/10.1063/1.4742167) which follows from Nanbu and Yonemura 1998 (http://dx.doi.org/10.1006/jcph.1998.6049). That method has been shown to result in incorrect relaxation rates for certain problems by Higginson et al 2020 (https://doi.org/10.1016/j.jcp.2020.109450). PR #5091 fixes this issue using a method similar to that by Higginson 2020 that produces correct relaxation rates with weighted particles. Below are results from Tests T1a-T1d from Higginson 2020 (but with 10X less particles). The physical setup is the same for each of tests T1a-T1d, but they use different particle weighting setups for the different populations. In contrast to results obtained using the development branch, the results obtained using this PR show identical relaxation results for the various particle weight combinations.

Using this PR:
T1abcd_binaryOpt_branch

Using the development branch (prior to PR #5091):
T1abcd_develop_branch_new

### CODE REVIEW

**Comment by RemiLehe** [comment]
Thanks a lot for this PR! For clarity and because this PR will probably serve as reference in the future by other WarpX developers, would it make sense to split this into 2 independent PR?
- One PR that fixes the performance issue, by computing the density and temperature ahead of time.
- One PR that fixes the weighting issue, by introducing this code https://github.com/ECP-WarpX/WarpX/pull/5047/files#diff-9a3d3b7442d673878318b7ebdf440a46b42570bb3afae0766af6c2301e910dcfR118 and the automated test

**Comment by JustinRayAngus** [comment]
I understand and agree that your suggestion is ideal. It was my original intent to do them separately. 

The issue is the n12 parameter computed in lines 99-108 of ElasticCollisionPerez.H. That value is an order N calc done for each binary pair (net cost is order N squared), but it's also not used at all in the improved method that produces the correct scattering. It seemed like a lot of unnecessary work to pull this parameter out to being computed at the cell level and then passing it down the chain of calls when it is not needed in the first place and all these added lines of code would just be removed in the follow on PR.

Unless you see a way to avoid the issue I described above, I think we should not split this PR into two. There will be a few more minor PRs related to Coulomb scattering following this one. 

What do you think? Perhaps I can update the PR title to better reflect both parts of this PR?

**Comment by RemiLehe** [comment]
OK, thanks a lot for clarifying.

How about switching the order of the PRs, i.e. doing a separate PR that implements only:

- One PR that fixes the weighting issue, by introducing this code https://github.com/ECP-WarpX/WarpX/pull/5047/files#diff-9a3d3b7442d673878318b7ebdf440a46b42570bb3afae0766af6c2301e910dcfR118 and the automated test

(that way, the calculation of `n12` is not needed in the code anymore)

and then once the above is merged, we could merge the current PR (i.e. #5047), which, at this point, will not show any changes for https://github.com/ECP-WarpX/WarpX/pull/5047/files#diff-9a3d3b7442d673878318b7ebdf440a46b42570bb3afae0766af6c2301e910dcfR118 since that part will already have been merged as part of the first PR.

Does that make sense, and does it sound realistic? If not, I agree we can also simply merge this PR (i.e. #5047) as a single combined PR.

**Comment by JustinRayAngus** [comment]
Splitting that way would be less work, but to be safe I should probably re-run all the tests (T1a-T1d), which will take a while because of the N^2 cost. If the aim of splitting the PRs in two is to be able to go back to the Nanbu method at some point for making comparisons or something, then I think it would make more sense to actually do it the first way you propose where the first PR is just about the optimization. That way also doesn't require re-running any of the new tests, since the code will fail for them anyhow.

For the sake of time and moving forward, I would advocate for simply merging this PR as is. This is my biased opinion. I'll let you have the final say of what to do.

**Comment by ax3l** [comment]
I like that split, that way we have a clear bug fix PR and a clear performance optimization PR. This will make also future potential regressions easier to spot and address (or reverts).

Our idea is not to go back to Nanbu in the future. It is more about having PRs that are easier to understand, to review and to go back to in the future.

File: Source/Particles/Collision/BinaryCollision/Coulomb/ElasticCollisionPerez.H
Comment (RemiLehe) [inline Source/Particles/Collision/BinaryCollision/Coulomb/ElasticCollisionPerez.H]: The factor `min_N` is not explained in Higginson et al. (since Higginson's paper is duplicating particles instead, which results in a factor `max_N` instead). 

My understanding is that, for a given species (say species 1), this factor should correspond to the number of macroparticles in the other species (`N2` for species 1), in order to account for the fact that a macroparticle from species 1 will only collide with one or (a few) macroparticles of species 2, instead of all of the `N2` macroparticles. 
If species 1 is the species with the larger number of macroparticles, this automatically works (`min_N = N2`).
If species 2 is the species with the lowest number of macroparticles, then its particles will participate, on average, in `max_N/min_N` pairs, so it is as if they see a factor `max_N/min_N x min_N = max_N = N2`, so it works in this case.

Is the above correct? If yes, could you add an additional comment explaining the factor `min_N` in the code (if possible, in a more concise and clearer way than the explanation I gave above).

Comment (RemiLehe) [reply Source/Particles/Collision/BinaryCollision/Coulomb/ElasticCollisionPerez.H]: Note also that the other binary collisions in WarpX use a duplication method similar to Higginsons' paper:

https://github.com/ECP-WarpX/WarpX/blob/development/Source/Particles/Collision/BinaryCollision/DSMC/DSMCFunc.H#L147
https://github.com/ECP-WarpX/WarpX/blob/development/Source/Particles/Collision/BinaryCollision/DSMC/DSMCFunc.H#L173

https://github.com/ECP-WarpX/WarpX/blob/development/Source/Particles/Collision/BinaryCollision/NuclearFusion/NuclearFusionFunc.H#L186
https://github.com/ECP-WarpX/WarpX/blob/development/Source/Particles/Collision/BinaryCollision/NuclearFusion/NuclearFusionFunc.H#L173

Thus, it might be more consistent to use Higginson's duplication method in the Coulomb module too.

Are there advantages to the proposed method without duplication, aside from the slightly simpler implementation?

File: Examples/Tests/collision/inputs_1d
Comment (RemiLehe) [inline Examples/Tests/collision/inputs_1d]: Why is this `epsilon` needed? Is this in the test described by Higginson?

File: Examples/Tests/collision/inputs_1d
Comment (RemiLehe) [inline Examples/Tests/collision/inputs_1d]:
```suggestion
```
Since the CI tests run only on 2 cores, it might be more efficient to use the default `max_grid_size`.

File: Examples/Tests/collision/analysis_collision_1d.py
Comment (RemiLehe) [inline Examples/Tests/collision/analysis_collision_1d.py]:
```suggestion
```

File: Examples/Tests/collision/analysis_collision_1d.py
Comment (RemiLehe) [inline Examples/Tests/collision/analysis_collision_1d.py]:
```suggestion
```

File: Examples/Tests/collision/analysis_collision_1d.py
Comment (RemiLehe) [inline Examples/Tests/collision/analysis_collision_1d.py]: I understand that this makes the script generic, but for a CI test, I think it is good to simplify it and keep only the code that will actually be run in the CI test, i.e. removing `test_tag` and the corresponding `if` conditions.

File: Examples/Tests/collision/analysis_collision_1d.py
Comment (RemiLehe) [inline Examples/Tests/collision/analysis_collision_1d.py]:
```suggestion
# Separate macroparticles from group A (low weight) and group B (high weight)
# by sorting based on the weight
```

File: Examples/Tests/collision/analysis_collision_1d.py
Comment (RemiLehe) [inline Examples/Tests/collision/analysis_collision_1d.py]:
```suggestion
# Find the index `Npmin` that separates macroparticles from group A and group B
# in the sorted arrays.
```

File: Examples/Tests/collision/analysis_collision_1d.py
Comment (RemiLehe) [inline Examples/Tests/collision/analysis_collision_1d.py]: Slight reformulation:
```suggestion
sorted_indices = data['ions','particle_weight'].argsort()
sorted_wp = data['ions','particle_weight'][sorted_indices].value
```

File: Examples/Tests/collision/analysis_collision_1d.py
Comment (RemiLehe) [inline Examples/Tests/collision/analysis_collision_1d.py]:
```suggestion
```

File: Examples/Tests/collision/analysis_collision_1d.py
Comment (RemiLehe) [inline Examples/Tests/collision/analysis_collision_1d.py]:
```suggestion
```

File: Examples/Tests/collision/analysis_collision_1d.py
Comment (RemiLehe) [inline Examples/Tests/collision/analysis_collision_1d.py]: Could you confirm that this fails when using the WarpX version from `development` branch (i.e. without this PR) ?
Also: would it make sense to also add an `assert` checking the other quantities (`TAx`, `TAy`, as well as the `TB`s, `uA`s and `uB`s)?
And would it be possible to run the simulation a bit longer (e.g. 100 ps) so that we can actually use the equilibrium values of the `T`s and `u`s, which can be computed analytically?

File: Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H
Comment (RemiLehe) [inline Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H:155]: Could add a reference in the comment? (Is it the same reference as above?)

File: Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H
Comment (RemiLehe) [inline Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H:144]: Given that we cite the above paper, I think that this is not needed.
```suggestion
```

File: Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H
Comment (RemiLehe) [inline Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H]:
```suggestion
    // Compute variant relative velocity in center-of-mass frame
```
Also: What is meant by `variant` here?

File: Examples/Tests/collision/inputs_1d
Comment (JustinRayAngus) [reply Examples/Tests/collision/inputs_1d]: That is there so that one can distinguish the two population of particles in the output file for test T1a when using equally-weighted particles. I didn't know how else to do it other than having one population with a slightly different weight than the other.

File: Examples/Tests/collision/analysis_collision_1d.py
Comment (JustinRayAngus) [reply Examples/Tests/collision/analysis_collision_1d.py]: I chose this time because this is where one sees the largest discrepency between the new corrected implementation and the original implementation. If you ran until equilibrium, then the development branch will not fail the check. It is the relaxation to the equilibrium that is incorrect. Not the steady state solution.

Comment (JustinRayAngus) [reply Examples/Tests/collision/analysis_collision_1d.py]: And yes. The development branch fails this test.

File: Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H
Comment (JustinRayAngus) [reply Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H]: It is to distinguish it from the invariant relative velocity defintion that is frame independent.

Comment (JustinRayAngus) [reply Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H]: Will make comment change.

File: Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H
Comment (JustinRayAngus) [reply Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H:144]: I felt that since the formulation here largely follows Perez 2012, that it may be useful to make explicit comments where things differ. But if you feel it is not needed then I will take it out.

File: Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H
Comment (JustinRayAngus) [reply Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H:155]: Using sigma_max is the same as using sp from Perez 2012. Did you get a chance to look at the tech-note I sent you and Axel where I went over the different limiting models used by Perez?

File: Examples/Tests/collision/analysis_collision_1d.py
Comment (JustinRayAngus) [reply Examples/Tests/collision/analysis_collision_1d.py]: Ok

File: Examples/Tests/collision/analysis_collision_1d.py
Comment (JustinRayAngus) [reply Examples/Tests/collision/analysis_collision_1d.py]: Ok

File: Examples/Tests/collision/analysis_collision_1d.py
Comment (JustinRayAngus) [reply Examples/Tests/collision/analysis_collision_1d.py]: Ok

File: Examples/Tests/collision/analysis_collision_1d.py
Comment (JustinRayAngus) [reply Examples/Tests/collision/analysis_collision_1d.py]: Ok

File: Examples/Tests/collision/analysis_collision_1d.py
Comment (JustinRayAngus) [reply Examples/Tests/collision/analysis_collision_1d.py]: Ok

File: Examples/Tests/collision/analysis_collision_1d.py
Comment (JustinRayAngus) [reply Examples/Tests/collision/analysis_collision_1d.py]: Ok

File: Examples/Tests/collision/analysis_collision_1d.py
Comment (JustinRayAngus) [reply Examples/Tests/collision/analysis_collision_1d.py]: Ok

File: Examples/Tests/collision/analysis_collision_1d.py
Comment (JustinRayAngus) [reply Examples/Tests/collision/analysis_collision_1d.py]: Ok

File: Examples/Tests/collision/inputs_1d
Comment (JustinRayAngus) [reply Examples/Tests/collision/inputs_1d]: Ok

File: Examples/Tests/collision/inputs_1d
Comment (JustinRayAngus) [reply Examples/Tests/collision/inputs_1d]: Since the CI test does not do test T1a. I can remove the epsilon.

File: Source/Particles/Collision/BinaryCollision/Coulomb/ElasticCollisionPerez.H
Comment (JustinRayAngus) [reply Source/Particles/Collision/BinaryCollision/Coulomb/ElasticCollisionPerez.H]: It is not just Nmax vs Nmin that is the difference. In the Higginson method, one uses (wmax_prime)*Nmax, where wmax_prime = max(wi/df,wj/df) is the max modified weight of the two particles with dp the duplication factor. Note that the average duplication factor per pair is Nmax/Nmin, which means on average (wmax_prime) *Nmax = wmax *Nmin.

Using Nmin is certainly simpler because you don't have to used a modified weight that takes into account a pair-specific duplication factor. It is also the most direct extension of the original TA77 method, which uses Nmin.

Comment (JustinRayAngus) [reply Source/Particles/Collision/BinaryCollision/Coulomb/ElasticCollisionPerez.H]: > The factor `min_N` is not explained in Higginson et al. (since Higginson's paper is duplicating particles instead, which results in a factor `max_N` instead).
> 
> My understanding is that, for a given species (say species 1), this factor should correspond to the number of macroparticles in the other species (`N2` for species 1), in order to account for the fact that a macroparticle from species 1 will only collide with one or (a few) macroparticles of species 2, instead of all of the `N2` macroparticles. If species 1 is the species with the larger number of macroparticles, this automatically works (`min_N = N2`). If species 2 is the species with the lowest number of macroparticles, then its particles will participate, on average, in `max_N/min_N` pairs, so it is as if they see a factor `max_N/min_N x min_N = max_N = N2`, so it works in this case.
> 
> Is the above correct? If yes, could you add an additional comment explaining the factor `min_N` in the code (if possible, in a more concise and clearer way than the explanation I gave above).

Your interpretation is correct. It is also exactly the same interpretation of the original method for equally-weighted particles by Takizuka and Abe 1977.

File: Examples/Tests/collision/inputs_1d
Comment (JustinRayAngus) [reply Examples/Tests/collision/inputs_1d]: Removing amr.blocking_factor leads to crash on run time.

File: Source/Particles/Collision/BinaryCollision/Coulomb/ElasticCollisionPerez.H
Comment (JustinRayAngus) [reply Source/Particles/Collision/BinaryCollision/Coulomb/ElasticCollisionPerez.H]: You asked about advantages of the proposed method.

As you stated, one advantage is that it is simpler. Rather than adding un-necessary lines of code here to maintain quasi-consistency with DSMC and Fusion methods, I would suggest considering adopting the simplified approach here in those modules. In addition to being simpler, there is in fact another advantage when using it for reactions that result in product species such as Fusion. I’ll try to concisely explain an issue with the current approach used in the fusion module below.

For simplicity, consider inter-species fusion between two species with equal weight particles w and using fusion_multiplier = 1. Using N_max, the reaction probability will be too large by N_max/N_min on average. However, because the reaction weight used is w_min = N_min/N_max*w (on average), the total number of fusion products produced is correct on average. So, the current method essentially uses a fusion_multiplier of N_max/N_min even when the specified fusion_multiplier is 1. Thus, it is not possible to do a simulation with equally weighted particles as the particles will no longer be equally weighted after a single fusion event when N1 and N2 are not equal. This approach can also result in the avalanching production of small weight particles described in PR #4976 

The alternative approach has an advantage in that it allows one to maintain equally-weighted particles and it will not have the issue described in PR #4976.

I'm happy to discuss further off line if needed.

**Comment by JustinRayAngus** [comment]
Looks like the checksum values for the new CI test are slightly different. Would the removal of max_grid_size line cause this?

File: Source/Particles/Collision/BinaryCollision/Coulomb/ElasticCollisionPerez.H
Comment (roelof-groenewald) [reply Source/Particles/Collision/BinaryCollision/Coulomb/ElasticCollisionPerez.H]: I agree we should switch the DSMC and fusion modules to also use this new method. Doing that for DSMC is on my to-do list.

**Comment by JustinRayAngus** [comment]
[CoulombScattering_s12.pdf](https://github.com/user-attachments/files/16555064/CoulombScattering_s12.pdf)

File: Regression/Checksum/benchmarks_json/collisionXYZ.json
Comment (JustinRayAngus) [inline Regression/Checksum/benchmarks_json/collisionXYZ.json]: The large difference in these temperatures is a bit concerning. I may need to inspect this test further.

Comment (JustinRayAngus) [reply Regression/Checksum/benchmarks_json/collisionXYZ.json]: Looks like this relatively large change happened on PR #5110.

File: Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H
Comment (ax3l) [reply Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H:144]: I am ok to leave it, explicit is good where things differ.

File: Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H
Comment (ax3l) [reply Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H:155]: This is great. If you like, the PDF of the simplification proof linked and the tech note could also be part of our [theory sections](https://warpx.readthedocs.io/en/latest/theory/intro.html). (Relatively quick to translate from LaTeX to `.rst`.)

**Review by ax3l — APPROVED** [review]
Thank you, this looks great to me.

Feel free to move some of the PDFs to the theory section (as .rst with `.. math::` blocks). That can be a follow-up, of course.

File: Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H
Comment (RemiLehe) [inline Source/Particles/Collision/BinaryCollision/Coulomb/UpdateMomentumPerezElastic.H]: Could you add a reference (e.g. pointing to a specific section of the Perez paper), so that future developers that read this understand where this comes from?

**Review by RemiLehe — APPROVED** [review]

**Pull request merged**
